OpenRA.Game/Sync.cs

```diff
@@ -10,158 +10,68 @@
-			{ typeof(Target), ((Func<Target, int>)HashTarget).Method },
-		};
-
-		static void EmitSyncOpcodes(Type type, ILGenerator il)
-		{
-			if (CustomHashFunctions.TryGetValue(type, out var hashFunction))
-				il.EmitCall(OpCodes.Call, hashFunction, null);
-			else if (type == typeof(bool))
-			{
-				var l = il.DefineLabel();
-				il.Emit(OpCodes.Ldc_I4, 0xaaa);
-				il.Emit(OpCodes.Brtrue, l);
-				il.Emit(OpCodes.Pop);
-				il.Emit(OpCodes.Ldc_I4, 0x555);
-				il.MarkLabel(l);
-			}
-			else if (type != typeof(int))
-				throw new NotImplementedException($"SyncAttribute on member of unhashable type: {type.FullName}");
-
-			il.Emit(OpCodes.Xor);
-		}
+		public static int Hash(CPos a) => a.Bits;
 
-		static Func<object, int> GenerateHashFunc(Type t)
-		{
-			var d = new DynamicMethod($"hash_{t.Name}", typeof(int), new Type[] { typeof(object) }, t);
-			var il = d.GetILGenerator();
-			var this_ = il.DeclareLocal(t).LocalIndex;
-			il.Emit(OpCodes.Ldarg_0);
-			il.Emit(OpCodes.Castclass, t);
-			il.Emit(OpCodes.Stloc, this_);
-			il.Emit(OpCodes.Ldc_I4_0);
-
-			const BindingFlags Binding = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;
-			foreach (var field in t.GetFields(Binding).Where(x => x.HasAttribute<SyncAttribute>()))
-			{
-				il.Emit(OpCodes.Ldloc, this_);
-				il.Emit(OpCodes.Ldfld, field);
-
-				EmitSyncOpcodes(field.FieldType, il);
-			}
+		public static int Hash(CVec a) => ((a.X * 5) ^ (a.Y * 3)) / 4;
 
-			foreach (var prop in t.GetProperties(Binding).Where(x => x.HasAttribute<SyncAttribute>()))
-			{
-				il.Emit(OpCodes.Ldloc, this_);
-				il.EmitCall(OpCodes.Call, prop.GetGetMethod(true), null);
+		public static int Hash(WDist a) => a.Length;
 
-				EmitSyncOpcodes(prop.PropertyType, il);
-			}
+		public static int Hash(WPos a) => a.X ^ a.Y ^ a.Z;
 
-			il.Emit(OpCodes.Ret);
-			return (Func<object, int>)d.CreateDelegate(typeof(Func<object, int>));
-		}
+		public static int Hash(WVec a) => a.X ^ a.Y ^ a.Z;
 
-		public static int HashInt2(int2 i2)
-		{
-			return ((i2.X * 5) ^ (i2.Y * 3)) / 4;
-		}
+		public static int Hash(WAngle a) => a.Angle;
 
-		public static int HashCPos(CPos i2)
-		{
-			return i2.Bits;
-		}
+		public static int Hash(WRot a) => Hash(a.Roll) ^ Hash(a.Pitch.Angle) ^ Hash(a.Yaw.Angle);
 
-		public static int HashCVec(CVec i2)
-		{
-			return ((i2.X * 5) ^ (i2.Y * 3)) / 4;
-		}
+		public static int Hash(Actor a) => a == null ? 0 : (int)(a.ActorID << 16);
 
-		public static int HashActor(Actor a)
-		{
-			if (a != null)
-				return (int)(a.ActorID << 16);
-			return 0;
-		}
+		public static int Hash(Player p) => p == null ? 0 : (int)(p.PlayerActor.ActorID << 16) * 0x567;
 
-		public static int HashPlayer(Player p)
-		{
-			if (p != null)
-				return (int)(p.PlayerActor.ActorID << 16) * 0x567;
-			return 0;
-		}
+		public static int Hash(object o) => throw new NotSupportedException($"Type {o.GetType().FullName} is not supported for Sync hashing!");
```

Comment:
I realised we might be able to bin this overload and move this to compile time if the source generator can detect that there is no matching overload and emit something like:
```c#
public virtual int GetSyncHash()
{
	#error "Type <foo> is not supported for Sync hashing on member <bar>"
	return 0;
}
```